### PR TITLE
Refactor code (disable Automod features)

### PR DESCRIPTION
**Pull request by ajayat**
Improve code style, update Docker configuration, and implement new moderation features including toxicity detection and message moderation based on optimized transformers. Update documentation.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -78,10 +81,17 @@ It will install packages in the virtual environment (recommended).
...
 
-- ### Using Docker
+### Using Docker
 
 ```sh
 docker compose up --build
 ```
+
+Now, the bot should be running. After modifying the code,
+you can update the bot using the command `docker-compose up`.
```
Comment (Alessevan) [inline README.md]:
```suggestion
you can update the bot using the command `docker compose up`.
```

**Review by Alessevan — CHANGES_REQUESTED** [review]
Tout d'abord, je ne suis pas certain de l'utilité d'une détection de la toxicité d'un message. Je pense que l'équipe de modération, à quatre, suffira amplement et dans le pire cas, il peut y avoir des mentions pour faire venir rapidement un membre de l'équipe du discord.
Les fichiers de plus de 100Mo sont-ils disponibles en ligne ? Le cas échéant, ils pourraient être téléchargés à l'exécution, cela réduirait la taille du dépôt et éviterait d'utiliser `git-lfs`. J'ai également un problème, sur cette branche, le dossier `.git/lfs` n'arrête pas de grossir à chaque `du -sh` à la racine du dépôt. Mais cela peut venir de moi, j'essaierai sur une autre PC demain.
Selon moi, la PR aurait pu être séparée en plusieurs vu tout ce qui est abordé au sein de celle-ci.
Enfin, il y a juste un point à changer.

**Comment by ajayat** [comment]
> Tout d'abord, je ne suis pas certain de l'utilité d'une détection de la toxicité d'un message. Je pense que l'équipe de modération, à quatre, suffira amplement et dans le pire cas, il peut y avoir des mentions pour faire venir rapidement un membre de l'équipe du discord.

C'est quand même plus sympa et pratique d'avoir une modération automatique de propos haineux sans avoir à attendre la disponibilité d'un modo (et ça évite que vous soyez ping tout le temps)

> Les fichiers de plus de 100Mo sont-ils disponibles en ligne ? Le cas échéant, ils pourraient être téléchargés à l'exécution, cela réduirait la taille du dépôt et éviterait d'utiliser git-lfs

Oui le modèle est disponible en ligne, mais j'ai optimisé le modèle en local puis exporté avec des outils qui peuvent être installés en faisant `pipenv install --dev`, et en utilisant la fonction `onnx_export_from_model` du module `utils/automod`. Le problème étant que ces modules font grossir le dépot de 250Mo à 1.2Go, et donc le temps initial de build de Docker.

> J'ai également un problème, sur cette branche, le dossier .git/lfs n'arrête pas de grossir à chaque du -sh à la racine du dépôt

C'est dû au fonctionnement de Git LFS qui met en cache les anciens fichiers suivis. Tu peux faire `git lfs prune` et le problème sera réglé !

> Selon moi, la PR aurait pu être séparée en plusieurs vu tout ce qui est abordé au sein de celle-ci.

Effectivement, j'aurai pu séparer la refactorisation et l'automod, fin bon c'est trop tard maintenant mdr

J'ai push le modèle `roberta` qui fait 500 Mo sur le disque mais 1050Mo dans la RAM car les float16 sont représentés par des float32 sur cpu. `distilbert` fait quant à lui 500 Mo dans la RAM.
Il est possible de changer le modèle dans `_classifier = ToxicityClassifier(MODEL_DIR / "<nom-du-modèle>")`, ensuite la pipeline est la même (il faudrait affiner le seuil suivant le modèle), il y a juste softmax/sigmoid qui sont utilisés suivant le nombre de classes pour calculer la probabilité de toxicité.

Niveau performances, je n'ai pas encore fait les test théoriques avec AUC-ROC, mais en pratique le modèle `roberta` est vraiment très précis et détecte bien les contenus inapropriés. Le modèle `distilbert` bien qu'il soit précis avec un faible FPR, il a du mal à détecter toutes les insultes en français.

Enfin, le bot aurait une consommation de 1Go sur le VPS, ce qui ne devrait pas être un problème, pour l'instant, en comptant les autres services. De manière générale, c'est dommage de limiter les fonctionnalités à cause de contraintes matérielles. On va certainement upgrader le VPS tôt ou tard, à mesure que l'association et les services évoluent.

**Comment by ajayat** [comment]
Enfin, quand j'aurai un peu de temps, j'essaierai de finetuner un modèle en me basant sur `distilbert-multilingual` sur ce dataset https://www.kaggle.com/competitions/jigsaw-toxic-comment-classification-challenge/data. Ça me paraît pas déconnant de pouvoir atteindre des performances équivalentes à `roberta` pour un modèle 2x plus léger

**Comment by ajayat** [comment]
J'ai optimisé le modèle `roberta` avec des poids sur 8 bits notamment, il est désormais 4x plus légers sur le disque (~250Mo) et consomme 700 Mo dans la RAM

File: mp2i/utils/automod.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+from mp2i.wrappers.guild import GuildWrapper
+
+logger = logging.getLogger(__name__)
+
+
+def is_toxic(msg: discord.Message) -> bool:
+    """
+    Check if a message is toxic or not.
+    """
+    return _classifier.predict(msg.content, treshold=0.9)
```
Comment (Triw-12) [inline mp2i/utils/automod.py]:
```suggestion
    return _classifier.predict(msg.content, threshold=0.9)
```
A remplacer aussi ligne 114 et 125

**Review by Triw-12 — CHANGES_REQUESTED** [review]
J'ai pas tout tester encore mais j'ai déjà quelques remarques à faire :grin: 

J'aurai aussi préféré que ce soit séparé en plusieurs PR mais pas grave :grin:  

Je ne suis toujours pas convaincu que l'IA va fonctionner mais je pense que on peut avoir une periode de test pour voir si : 1. ça tient sur le VPS, 2. c'est utile et ça modère bien les messages.

Concernant l'évolution du VPS, je pense qu'il faudrait plutôt en parler dans le salon gestion mais perso je trouve que 30€ par an pour un automod c'est un peu cher :grin: (y'a pas que ça mais on peut en parler dans gestion)

Sinon, ce serait bien de rajouter le fait que **l'auteur du message reçoive un message** disant que son message a été automod pour éviter les incompréhensions.

Rajouter dans le README.md le fait qu'il faut utiliser la commande `git lfs pull`

**Comment by ajayat** [comment]
Je suis convaincu que ça va fonctionner, mais on peut faire les tests sur le VPS, j'espère que ça pourra te convaincre ! 

> l'auteur du message reçoive un message 

En effet, ça peut être bien ! je fais ça.

Pour Git LFS, normalement en faisant `git lfs install` au début, `git pull` va automatiquement récupérer les fichiers lfs ? (à vérifier)

File: mp2i/utils/automod.py (hunk 2 above)
Comment (Triw-12) [reply mp2i/utils/automod.py]: Si tu rechanges la modification dans le commit d'après ça va pas marcher :joy: 
Faut aussi modifier les lignes 144 et 125

**Comment by ajayat** [comment]
J'ai déplacé l'automod dans un Cog (module) dédié, et renommé l'utilitaire `utils/classifier.py`
Enfin, j'ai créé une commande `exception` pour ajouter une règle du type `pattern -> remplacement` afin d'ajouter des exceptions à l'automod. Pour l'instant c'est sauvegardé dans un json mais on pourrait peut-être le faire dans une table si vous trouvez ça mieux (après ce n'est pas des données privées, ça peut faire "partie" de l'automod)

File: models/multilingual-toxic-xlm-roberta/config.json [hunk 3]
```diff
@@ -0,0 +1,36 @@
+{
+  "_attn_implementation_autoset": true,
+  "_name_or_path": "/home/adrien/Documents/Programmation/Github/mp2i-discord-bot/models/multilingual-toxic-xlm-roberta",
```
Comment (Triw-12) [inline models/multilingual-toxic-xlm-roberta/config.json]: Je crois que ton chemin absolu du model n'a rien à faire là

File: mp2i/cogs/automod.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            return json.load(f)
+
+    def is_toxic(self, msg: discord.Message) -> bool:
+        """
+        Check if a message is toxic or not.
+        """
+        for exc in self.exceptions:
+            msg.content = re.sub(exc, self.exceptions[exc], msg.content)
+
+        return self.classifier.predict(msg.content, threshold=0.75)
```
Comment (Triw-12) [inline mp2i/cogs/automod.py]: Comme c'est une variable importante, ce serait pas mieux de le mettre en variable globale pour le mettre en évidence que au milieu du fichier ?

File: mp2i/cogs/suggestions.py [hunk 5]
```diff
@@ -171,48 +180,49 @@ async def suggestions(self, ctx, state: str) -> None:
...
-                embed.add_field(
-                    name=f"{i+1} - Suggestion de {user.name if user else 'Utilisateur inconnu'} le {suggestion.date:%d/%m/%Y}",
-                    value=message.jump_url,
-                    inline=False,
-                )
+                suggestion_channel = guild.suggestion_channel
+                try:
+                    msg = await suggestion_channel.fetch_message(suggest.message_id)
+                except discord.errors.NotFound:
+                    continue
```
Comment (Triw-12) [inline mp2i/cogs/suggestions.py:215]: On avait pas dit que on loguer une erreur du coup ?

File: mp2i/cogs/automod.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    @guild_only()
+    @hybrid_command(name="exception")
+    async def add_exception(self, ctx, pattern: str, replacement: str) -> None:
+        """
+        Add a pattern to the exceptions dictionary.
+        """
+        self.exceptions[pattern] = replacement
+        with open(STATIC_DIR / "text/exceptions.json", "w") as f:
+            json.dump(self.exceptions, f, indent=2)
+
```
Comment (Triw-12) [inline mp2i/cogs/automod.py:110]: Je ne comprends vraiment pas ta solution. Tu veux avoir quelque chose de permanent sans le mettre dans la base de données... Le fichier n'est même pas dans le .gitignore :sweat_smile: 
Mais surtout, je ne comprends vraiment pas pourquoi tu fais quelque chose qui devrait être géré par la base de données si l'on veut quelque chose de persistant, pouvant être modifié et supprimé facilement... Parce que techniquement, si l'on veut supprimer/modifier une ligne, il faut se connecter au VPS, aller sur la bonne ligne, la modifier/supprimer, puis relancer le bot, au lieu de passer par pgAdmin, par exemple.
D'ailleurs, cela signifie qu'on ne peut pas retirer ou modifier facilement une ligne avec une commande, ce qui est assez dérangeant si l'on a fait une erreur ou si le mot de remplacement n'est pas suffisamment "safe".

Et pour l’instant, il n’y a aucune protection si l’on ajoute deux fois le même mot à remplacer...

**Review by Triw-12 — CHANGES_REQUESTED** [review]

File: models/multilingual-toxic-xlm-roberta/config.json (hunk 3 above)
Comment (ajayat) [reply models/multilingual-toxic-xlm-roberta/config.json]: En effet ! en fait ce fichier sert à rien, il est juste là pour "décrire" le modèle, donc je pense je vais le supprimer

File: mp2i/cogs/automod.py (hunk 4 above)
Comment (ajayat) [reply mp2i/cogs/automod.py]: On peut mixer les deux, mais c'est bien de garder de la flexibilité si par exemple on veut un seuil différent suivant les taches ou les choses qu'on veut modérer.
Mais pour la classe Automod je peux en effet mettre un seuil par défaut en tant qu'attribut de classe.

File: mp2i/cogs/suggestions.py (hunk 5 above)
Comment (ajayat) [reply mp2i/cogs/suggestions.py:215]: Non, tu confonds avec les roles je crois. Ici on peut mais en vrai un peu osef pour une suggestion

File: mp2i/cogs/automod.py (hunk 6 above)
Comment (ajayat) [reply mp2i/cogs/automod.py:110]: Déja répondu ici https://github.com/prepas-mp2i/mp2i-discord-bot/pull/22#issuecomment-2705452744

Pour modifier une ligne il suffit simplement d'ajouter une règle avec le même pattern, comme c'est un dictionnaire.
Pour supprimer, il suffit d'ajouter la règle `mot -> mot` , ou je peux ajouter une commande pour suppriemr réellement la ligne.

Comment (ajayat) [reply mp2i/cogs/automod.py:110]: Je me répète mais il est tout à fait possible de faire une table, ça ne me dérange pas non plus c'est aussi une solution. Mais quels sont les réels problèmes d'avoir un fichier ici ? Veut on garder ces données confidentielles ?

Comment (Triw-12) [reply mp2i/cogs/automod.py:110]: En fait, le principal problème est que le fichier n'est pas facilement accessible si on veut faire des modifications rapides et des corrections. Ce qui est "moins" le cas avec une base de donnée
De plus, si il est pas dans le .gitignore le bot peut modifier un fichiers dans la branche master, donc ça peut en soit créer des conflits si on modifie en parallèle ce fichier dans la branche develop puis que on met sur master

**Comment by Alessevan** [comment]
La fonctionnalité majeure de cette PR est rejettée par majorité de l'administration, l'ensemble de l'équipe de modération et la plupart (pour ne pas dire tous) des membres actifs.
